src/modm/platform/uart/rtt/rtt.cpp.in

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2021, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <modm/platform/device.hpp>
+#include "rtt.hpp"
+#include <algorithm>
+
+namespace modm::platform
+{
+
+struct RttBuffer
+{
+	const char* name{nullptr};
+	uint8_t* buffer{nullptr};
+	const uint32_t size{0};
+	volatile uint32_t head{0};
+	volatile uint32_t tail{0};
+	uint32_t flags{0};
+
+	bool write(uint8_t data)
+	{
+		const uint32_t rhead{head};
+		const uint32_t rtail{tail};
+		const uint32_t rhead_next{(rhead + 1) % size};
+		if (not size or rhead_next == rtail) return false;
+		buffer[rhead] = data;
+		head = rhead_next;
+		return true;
+	}
+	bool read(uint8_t &data)
+	{
+		const uint32_t rhead{head};
+		const uint32_t rtail{tail};
+		if (not size or rtail == rhead) return false;
+		data = buffer[rtail];
+		tail = (rtail + 1) % size;
+		return true;
+	}
+	bool isEmpty() const { return (head == tail); }
+	bool getSize() const
+	{
+		const uint32_t rhead{head};
+		const uint32_t rtail{tail};
+		return ((rhead >= rtail) ? 0 : (size + 1)) + rhead - rtail;
```

Comment:
This doesn't look right. Shouldn't it be
```suggestion
		return ((rhead >= rtail) ? (rhead - rtail) : (size + rhead - rtail);
```
?

Reply:
I copied this from modm::atomic::Queue, but that leaves one element empty, which the RTT ring buffers does not. We could use better ring buffers…